X-Ray segments and sub-segments need to be identified clearly

`webapp` is giving the following error:

```
2018-03-27 17:31:52,093 ERROR [io.undertow.request] (default task-1) UT005023: Exception handling request to /: org.jboss.resteasy.spi.UnhandledException: com.amazonaws.xray.exceptions.SegmentNotFoundException: Failed to begin subsegment named 'get': segment cannot be found.
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:78)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:222)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:179)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:422)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:213)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:228)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:56)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:51)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:85)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.swarm.generated.FaviconErrorHandler.handleRequest(FaviconErrorHandler.java:61)
	at io.undertow.server.handlers.PathHandler.handleRequest(PathHandler.java:94)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:131)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1508)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:326)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:812)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.amazonaws.xray.exceptions.SegmentNotFoundException: Failed to begin subsegment named 'get': segment cannot be found.
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.amazonaws.xray.strategy.RuntimeErrorContextMissingStrategy.contextMissing(RuntimeErrorContextMissingStrategy.java:17)
	at com.amazonaws.xray.contexts.ThreadLocalSegmentContext.beginSubsegment(ThreadLocalSegmentContext.java:23)
	at com.amazonaws.xray.AWSXRayRecorder.beginSubsegment(AWSXRayRecorder.java:423)
	at com.amazonaws.xray.AWSXRay.beginSubsegment(AWSXRay.java:92)
	at org.aws.samples.compute.webapp.WebappController.getMessage(WebappController.java:25)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:140)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:295)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:249)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:236)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:406)
	... 43 more
```

Its not clear if we should be using segments or sub-segments. Also, their names are not clear.

--- a/services/greeting/src/main/java/org/aws/samples/compute/greeting/GreetingEndpoint.java
+++ b/services/greeting/src/main/java/org/aws/samples/compute/greeting/GreetingEndpoint.java
@@ -1,8 +1,8 @@
 package org.aws.samples.compute.greeting;
 
 import com.amazonaws.xray.AWSXRay;
-import com.amazonaws.xray.AWSXRayRecorder;
-import com.amazonaws.xray.entities.Segment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import javax.ws.rs.GET;
 import javax.ws.rs.Path;
@@ -15,14 +15,15 @@ import javax.ws.rs.core.MediaType;
 @Path("greeting")
 public class GreetingEndpoint {
 
+    private static final Logger logger = LoggerFactory.getLogger(GreetingEndpoint.class);
+
     @GET
     @Produces(MediaType.TEXT_PLAIN)
     public String get() {
-//        AWSXRayRecorder xrayRecorder = AWSXRay.getGlobalRecorder();
-//        Segment segment = xrayRecorder.beginSegment("greeting");
-//        segment.putAnnotation("parentId", xrayRecorder.getTraceEntity().getId());
+        logger.info("get");
+        AWSXRay.getCurrentSegment().putAnnotation("parentId",
+                AWSXRay.getGlobalRecorder().getTraceEntity().getId());
         String response = "Hello";
-//        xrayRecorder.endSegment();
 
         return response;
     }

--- a/services/greeting/src/main/java/org/aws/samples/compute/greeting/MyApplication.java
+++ b/services/greeting/src/main/java/org/aws/samples/compute/greeting/MyApplication.java
@@ -1,5 +1,8 @@
 package org.aws.samples.compute.greeting;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import java.util.Set;
 
 import javax.ws.rs.ApplicationPath;
@@ -14,18 +17,21 @@ import static org.aws.samples.compute.greeting.MyApplication.APP_ROOT;
 public class MyApplication extends Application {
 
     public static final String APP_ROOT = "/resources";
+    private static final Logger logger = LoggerFactory.getLogger(MyApplication.class);
 
-//    @Override
-//    public Set<Class<?>> getClasses() {
-//        Set<Class<?>> classes = new java.util.HashSet<>();
-//        classes.add(GreetingEndpoint.class);
-//        return classes;
-//    }
-//
-//    @Override
-//    public Set<Object> getSingletons() {
-//        Set<Object> resources = new java.util.HashSet<>();
-//        resources.add(StartupBean.getInstance());
-//        return resources;
-//    }
+    @Override
+    public Set<Class<?>> getClasses() {
+        logger.info("getClasses");
+        Set<Class<?>> classes = new java.util.HashSet<>();
+        classes.add(GreetingEndpoint.class);
+        return classes;
+    }
+
+    @Override
+    public Set<Object> getSingletons() {
+        logger.info("getSingletons");
+        Set<Object> resources = new java.util.HashSet<>();
+        resources.add(StartupBean.getInstance());
+        return resources;
+    }
 }

--- a/services/name/src/main/java/org/aws/samples/compute/name/MyApplication.java
+++ b/services/name/src/main/java/org/aws/samples/compute/name/MyApplication.java
@@ -25,7 +25,7 @@ public class MyApplication extends Application {
     @Override
     public Set<Object> getSingletons() {
         Set<Object> resources = new java.util.HashSet<>();
-//        resources.add(StartupBean.getInstance());
+        resources.add(StartupBean.getInstance());
         return resources;
     }
 }

--- a/services/name/src/main/java/org/aws/samples/compute/name/NameEndpoint.java
+++ b/services/name/src/main/java/org/aws/samples/compute/name/NameEndpoint.java
@@ -1,8 +1,6 @@
 package org.aws.samples.compute.name;
 
 import com.amazonaws.xray.AWSXRay;
-import com.amazonaws.xray.AWSXRayRecorder;
-import com.amazonaws.xray.entities.Segment;
 
 import javax.ws.rs.GET;
 import javax.ws.rs.Path;
@@ -19,10 +17,9 @@ public class NameEndpoint {
     @Produces({MediaType.APPLICATION_XML + "; qs=0.50",
             MediaType.APPLICATION_JSON + "; qs=0.75"})
     public Name[] get() {
-//        AWSXRayRecorder xrayRecorder = AWSXRay.getGlobalRecorder();
-//        Segment segment = xrayRecorder.beginSegment("name");
-//        segment.putAnnotation("parentId", xrayRecorder.getTraceEntity().getId());
-//        xrayRecorder.endSegment();
+        AWSXRay.getCurrentSegment().putAnnotation("parentId",
+            AWSXRay.getGlobalRecorder().getTraceEntity().getId());
+
         return Names.findAll();
     }
     
@@ -32,10 +29,9 @@ public class NameEndpoint {
             MediaType.APPLICATION_JSON + "; qs=0.75",
             MediaType.TEXT_PLAIN + "; qs=1.0"})
     public Name get(@PathParam("id") int id) {
-//        AWSXRayRecorder xrayRecorder = AWSXRay.getGlobalRecorder();
-//        Segment segment = xrayRecorder.beginSegment("name1");
-//        segment.putAnnotation("parentId", xrayRecorder.getTraceEntity().getId());
-//        xrayRecorder.endSegment();
+        AWSXRay.getCurrentSegment().putAnnotation("parentId",
+            AWSXRay.getGlobalRecorder().getTraceEntity().getId());
+
         return Names.findById(id);
     }
 }

--- a/services/webapp/src/main/java/org/aws/samples/compute/webapp/WebappController.java
+++ b/services/webapp/src/main/java/org/aws/samples/compute/webapp/WebappController.java
@@ -2,8 +2,10 @@ package org.aws.samples.compute.webapp;
 
 import com.amazonaws.xray.AWSXRay;
 import com.amazonaws.xray.AWSXRayRecorder;
+import com.amazonaws.xray.entities.Namespace;
 import com.amazonaws.xray.entities.Segment;
 import com.amazonaws.xray.entities.Subsegment;
+import com.amazonaws.xray.entities.TraceHeader;
 import com.mashape.unirest.http.Unirest;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -24,40 +26,48 @@ public class WebappController {
     @Produces(MediaType.TEXT_PLAIN)
     @GET
     public String getMessage(@Context UriInfo uri, @PathParam("id") String id) {
-//        AWSXRayRecorder xrayRecorder = AWSXRay.getGlobalRecorder();
-//        Segment segment = xrayRecorder.beginSegment("webapp");
-//        segment.putAnnotation("parentId", xrayRecorder.getTraceEntity().getId());
-
         String greetingEndpoint = getEndpoint("GREETING", uri.getRequestUri().getScheme(), null);
         logger.info("ID Query is: " + id);
         String pathQuery = (id.equals("")) ? null : ("/" + id);
         String nameEndpoint = getEndpoint("NAME", uri.getRequestUri().getScheme(), pathQuery);
+
+        Segment segment = AWSXRay.getCurrentSegment();
+        AWSXRayRecorder xrayRecorder = AWSXRay.getGlobalRecorder();
+        segment.putAnnotation("parentId", xrayRecorder.getTraceEntity().getId());
+        Subsegment subsegment = xrayRecorder.beginSubsegment("greeting");
+        subsegment.setNamespace(Namespace.REMOTE.toString());
 
         String greetingMessage = "";
         try {
             greetingMessage = Unirest
                     .get(greetingEndpoint)
                     .header("accept", "text/plain")
+                    .header("x-amzn-trace-id", getTraceHeader(segment, subsegment).toString())
                     .asString()
                     .getBody();
             logger.info("Greeting is: " + greetingMessage);
         } catch (Exception e) {
             logger.error("Failed connecting Greeting API: " + e);
         }
+        xrayRecorder.endSubsegment();
+
+        subsegment = xrayRecorder.beginSubsegment("name");
+        subsegment.setNamespace(Namespace.REMOTE.toString());
 
         String nameMessage = "";
         try {
             nameMessage = Unirest
                     .get(nameEndpoint)
                     .header("accept", "text/plain")
+                    .header("x-amzn-trace-id", getTraceHeader(segment, subsegment).toString())
                     .asString()
                     .getBody();
             logger.info("Name is: " + nameMessage);
         } catch (Exception e) {
             logger.error("Failed connecting Name API: " + e);
         }
+        xrayRecorder.endSubsegment();
 
-//        xrayRecorder.endSegment();
         return greetingMessage + " " + nameMessage;
     }
 
@@ -96,6 +106,15 @@ public class WebappController {
 
         logger.info(type + " endpoint: " + endpoint);
         return endpoint;
+    }
+
+    private TraceHeader getTraceHeader(Segment segment, Subsegment subsegment) {
+        if (segment == null || subsegment == null)
+            return new TraceHeader();
+
+        return new TraceHeader(segment.getTraceId(),
+                segment.isSampled() ? subsegment.getId() : null,
+                segment.isSampled() ? TraceHeader.SampleDecision.SAMPLED : TraceHeader.SampleDecision.NOT_SAMPLED);
     }
 
 }

